services/src/main/java/org/keycloak/exportimport/ExportImportManager.java

```diff
@@ -95,6 +108,57 @@ public void runImport() {
         }
     }
 
+    public void runImportAtStartup(String dir, Strategy strategy) throws IOException {
+        ExportImportConfig.setReplacePlaceholders(true);
+        ExportImportConfig.setAction("import");
+
+        Stream<ProviderFactory> factories = sessionFactory.getProviderFactoriesStream(ImportProvider.class);
+
+        for (ProviderFactory factory : factories.collect(Collectors.toList())) {
+            String providerId = factory.getId();
+
+            if ("dir".equals(providerId)) {
+                ExportImportConfig.setDir(dir);
+                ImportProvider importProvider = session.getProvider(ImportProvider.class, providerId);
+                importProvider.importModel(sessionFactory, strategy);
+            } else if ("singleFile".equals(providerId)) {
```

Comment:
Yeah, we are running all available import providers. Right now we have two: `dir` and `singleFile`.

We need to run both in order to be able to keep the same experience when exporting using a single file or to a directory. Note, however, that even though we iterate over both we filter out the files that should be imported by the `singleFile` provider in order to avoid processing files twice. That means:

* Files using `-realm.json` prefix and containing `-users-` (e.g.: when exporting users in a separate file) are processed only by the `dir` provider.
* Any other file is processed by the `singleFile` provider

I would like to refactor the import providers a bit to make them less dependent on system properties and remove some code from here like those calling methods on `ExportImportConfig`, but I wanted to focus on the core issue.

If we want, we could also allow people to add their own import providers and change how import is done at startup time. But I don't have a good use case for it now.

Reply:
Thanks for the comment! It's more clear to me now.

> I would like to refactor the import providers a bit to make them less dependent on system properties and remove some code from here like those calling methods on ExportImportConfig, but I wanted to focus on the core issue.

Yep, it'd be great. The ExportImportConfig class is not very good :)) 

> If we want, we could also allow people to add their own import providers and change how import is done at startup time. But I don't have a good use case for it now.

IMHO, it could be a good improvement to allow add their import providers. And wouldn't be so time-consuming to implement it. Maybe it could also slightly polish the code.

Reply:
Not sure, import provider SPI is private. We don't want people using it and complaining about changes we made in the future ...

Let's keep an eye on this use case, if we find a strong argument for it, we do it.

Reply:
Ahhh, okay. Agree :))